packages/eslint-plugin/src/rules/prefer-optional-chain-utils/checkNullishAndReport.ts

```diff
@@ -0,0 +1,38 @@
+import { isTypeFlagSet } from '@typescript-eslint/type-utils';
+import type {
+  ParserServicesWithTypeInformation,
+  TSESTree,
+} from '@typescript-eslint/utils';
+import type {
+  ReportDescriptor,
+  RuleContext,
+} from '@typescript-eslint/utils/ts-eslint';
+import { unionTypeParts } from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import type {
+  PreferOptionalChainMessageIds,
+  PreferOptionalChainOptions,
+} from './PreferOptionalChainOptions';
+
+export function checkNullishAndReport(
+  context: RuleContext<
+    PreferOptionalChainMessageIds,
+    [PreferOptionalChainOptions]
+  >,
+  parserServices: ParserServicesWithTypeInformation,
+  { requireNullish }: PreferOptionalChainOptions,
+  maybeNullishNodes: TSESTree.Expression[],
+  descriptor: ReportDescriptor<PreferOptionalChainMessageIds>,
+): void {
+  if (
+    !requireNullish ||
+    maybeNullishNodes.some(node =>
+      unionTypeParts(parserServices.getTypeAtLocation(node)).some(t =>
+        isTypeFlagSet(t, ts.TypeFlags.Null | ts.TypeFlags.Undefined),
+      ),
+    )
```

Comment:
[Bug] I think this is the wrong behavior here:

```ts
declare const foo1: { bar: string | null };
foo1 && foo1.bar; // reported - not ok

declare const foo2: { bar: string };
foo2 && foo2.bar; // not reported - ok
```

[playground](https://deploy-preview-8559--typescript-eslint.netlify.app/play/#ts=5.4.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgjAFzwDe8ARrMZjAJYoDm8APvCgK4QTwC%2BA3AChcBeADJROPPgB0lGH3gB6RfDgAHJDAwhg8ALRskWJAGsBA0JFgJk6LMIBMxMnOoY6jXoMdiJj2bAKyoZY6praugamQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5ujABtcNhy9o-aJAA0ylVhXYYiAI6xKnAHIJGyKfPxxEu7AF9nAXWVuXQA&tsconfig=&tokens=false)

The rule shouldn't report in both cases, because the left part (`foo1 &&` and `foo2 &&`) is not nullish. The only difference is the type of the `bar` (nullish/non-nullish).

I think that the type of the last element of the chain should not be taken into account in this check.

Reply:
Good catch!